### PR TITLE
Updating to version 13

### LINKED ISSUE #569: Per-user authentication: Pass associated_user Auth Info to ShopifyAPI::Session
[`associated_user`](https://github.com/Shopify/shopify_app/blob/76c23425b9dcb1eb0cd6918fde1ef32bb05310e5/app/controllers/shopify_app/sessions_controller.rb#L57) should be [passed to `ShopifyAPI::Session.new`](https://github.com/Shopify/shopify_app/blob/76c23425b9dcb1eb0cd6918fde1ef32bb05310e5/app/controllers/shopify_app/sessions_controller.rb#L50) as [its 3rd argument](http://www.rubydoc.info/gems/shopify_api/ShopifyAPI%2FSession:initialize). This way, one can persist email, locale, etc... when storing a session when using `per_user_permissions`. 

This is what `ShopifyAPI::Session.new`'s 3rd argument is for.

### LINKED ISSUE #618: [Docs request] Add concurrent access mode examples (offline & online auth)
### Overview
Shopify's [OAuth API](https://help.shopify.com/en/api/getting-started/authentication/oauth/api-access-modes) and [Omniauth gem](https://github.com/Shopify/omniauth-shopify-oauth2) supports both offline and online access modes. We should provide a clear example of how to use them both, concurrently, in our apps. More specifically, how to setup online access for only the `ShopifyApp::AuthenticatedController` while still maintaining offline access for `ShopifyApp::AppProxyVerification` and `ShopifyApp::WebhookVerification`.

### Previous discussions
This has been discussed before, see #580, but there has not been a clear solution moving forward. As of today, there aren't any [public repos in Github implementing this](https://github.com/search?q=OmniAuth+Builder+provider+shopify+per_user_permissions+language%3Aruby&type=Code) either. From what I understood, you can build two Shopify tokens, one for offline and one for online, but I'm not sure how to scope them properly.

### Examples

Here's an example of what I'm talking about:

#### Offline setup (default) <-- This works
AuthenticatedControllers, proxy app, and webhooks work properly, but you can't access individual user info on AuthenticatedControllers.
```ruby
# config/omniauth.rb
Rails.application.config.middleware.use OmniAuth::Builder do
  # Offline access
  provider :shopify,
  ShopifyApp.configuration.api_key,
  ShopifyApp.configuration.secret,
  scope: ShopifyApp.configuration.scope,
  setup: lambda { |env|
    strategy = env['omniauth.strategy']

    shopify_auth_params = strategy.session['shopify.omniauth_params']&.with_indifferent_access
    shop = if shopify_auth_params.present?
      "https://#{shopify_auth_params[:shop]}"
    else
      ''
    end

    strategy.options[:client_options][:site] = shop
  }
end
```

#### Online setup <-- This works
AuthenticatedControllers work properly, but proxy app and webhooks are `Unauthorized`.
```ruby
# config/omniauth.rb
Rails.application.config.middleware.use OmniAuth::Builder do
  # Online access
  provider :shopify,
  ShopifyApp.configuration.api_key,
  ShopifyApp.configuration.secret,
  scope: ShopifyApp.configuration.scope,
  per_user_permissions: true, # <-- this changed
  setup: lambda { |env|
    strategy = env['omniauth.strategy']

    shopify_auth_params = strategy.session['shopify.omniauth_params']&.with_indifferent_access
    shop = if shopify_auth_params.present?
      "https://#{shopify_auth_params[:shop]}"
    else
      ''
    end

    strategy.options[:client_options][:site] = shop
  }
end
```

#### Concurrent setup (offline & online) <-- How does this work?
Expected: AuthenticatedControllers, proxy app, and webhooks work properly, and you can access the Shopify user info on AuthenticatedControllers.
```ruby
# config/omniauth.rb
Rails.application.config.middleware.use OmniAuth::Builder do
  # Offline access
  provider :shopify,
  ShopifyApp.configuration.api_key,
  ShopifyApp.configuration.secret,
  scope: ShopifyApp.configuration.scope,
  setup: lambda { |env|
    strategy = env['omniauth.strategy']

    shopify_auth_params = strategy.session['shopify.omniauth_params']&.with_indifferent_access
    shop = if shopify_auth_params.present?
      "https://#{shopify_auth_params[:shop]}"
    else
      ''
    end

    strategy.options[:client_options][:site] = shop
  }

  # Online access
  provider :shopify,
  ShopifyApp.configuration.api_key,
  ShopifyApp.configuration.secret,
  scope: ShopifyApp.configuration.scope,
  per_user_permissions: true,
  setup: lambda { |env|
    # ??? <-- Would this be different than the one above?
  }
end
```

Does this make sense? Or am I missing something obvious?

If someone has solved this in a way that is compatible with this gem, could you share it here? I'm willing to write the proper documentation once we've figured this out.

### Things I've learned so far

1. You can generate a new session using the [Shopify API](https://github.com/Shopify/shopify_api#getting-started) directly. Could adding an `around_action` to the AuthenticatedController that finds or creates a new online access session work? Instead of using the Omniauth::Builder?
2. We need to toggle between the offline and online token headers depending on what kind of access a certain action requires. As seen on [Shopify's OAuth instructions](https://help.shopify.com/en/api/getting-started/authentication/oauth#step-4-making-authenticated-requests) I believe we would have to set the proper token on each request.

### Related Notes

IMO concurrent access should be the default behavior for this gem. It's one less configuration step, shouldn't break current implementations, is more secure for embedded apps, and would allow use of proxy apps, webhooks, and embedded apps with users (together) right from the get-go, which I feel is a common, and amazing, combination for apps.

### Next steps
I'm opening this issue to see if I'm missing something before working on a potential solution. And if the feedback is positive, to see if anyone would like to help.

Thank you ❤️

### LINKED ISSUE #883: Per User tokens being used for Webhooks Callback Controller
Currently, after the release of the per_user_authentication, there is an issue with callbacks/webhooks where the user token is currently used. This will work for a time but the user tokens have a short expiry.

To fix this we need to 
- Modify our installation process to get both tokens
- Store both shop and user tokens
- Look up shop token in callbacks controller using shop domain
- [Maybe] Add an `app/uninstall` webhook required so that we can track shop session to require re-auth
- [Maybe] Add a way to check if shop token is still valid so that we can skip shop auth, so that shop auth is not repeated every time the app is used

@slucaskim

### PR DESCRIPTION
**Pull request by tanema**
This adds a migration guide, adds to the change log an updates the version

fixes #883
fixes #569
fixes #618

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,8 @@
+13.0.0
+------
++ Added concurrent user and shop session support (online/offline)
```
Comment (katiedavis) [inline CHANGELOG.md]: Wanna add PR number? Just for folks in the future

**Comment by rezaansyed** [comment]
@tanema @slucaskim: @ragalie and I found a bug where in shopify where it wasn'y handling nil `per_user_permissions` correctly. We have a fix up: https://github.com/Shopify/omniauth-shopify-oauth2/pull/88 . Recommendation would be to hold off on merging until we have `omniauth-shopify-oauth2` merged.

**Comment by tanema** [comment]
Okay I will wait for omniauth to be updated. That PR looks good. We can also update ShopifyAPI when we update that as well.

**Comment by rezaansyed** [comment]
@tanema and @slucaskim : the PR for handling `nil` `per_user_permissions` has been merged and deployed. Should be good to merge now.

**Comment by tanema** [comment]
Okay I updated omniauth and shopify_api. I gave it a quick tophat and everything seems good. I am shipping this today. Please keep an eye out to help out with any support on this in the near future.

**Pull request merged**